### PR TITLE
[sil-escape-analysis] Various improvements

### PR DESCRIPTION
**Pull request by swiftix**
Tech escape analysis a few new tricks:
- It is now able to look through basic block arguments
- It is able to analyze destructors of local objects

For more detailed description see the specific commits.

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by swiftix** [comment]
@eeckstein Erik, do you mind reviewing this PR? I suggest looking at separate commits, because it is easier to understand what is going on then.

**Comment by swiftix** [comment]
@swift-ci Please smoke test Linux

**Review by eeckstein — CHANGES_REQUESTED** [review]
LTGM!
But please add more test(s) to check the destructor analysis.

Thanks for doing this work!

File: test/SILOptimizer/escape_analysis.sil [hunk 1]
```diff
@@ -1310,3 +1311,91 @@ bb(%0 : $X, %1: $@callee_owned (X, X) -> (@out X, @error ErrorType)):
...
+
+bb2:
+  %5 = alloc_ref $X
+  br bb3(%5 : $X)
+
+bb3(%8: $X):
+  strong_release %8 : $X
+  %9 = tuple ()
+  return %9 : $()
+}
```
Comment (swiftix) [inline test/SILOptimizer/escape_analysis.sil]: @eeckstein I added a couple of tests to demonstrate that destructors which allow objects to escape are handled properly.

**Review by eeckstein — APPROVED** [review]

**Comment by swiftix** [comment]
@swift-ci Please smoke test and merge

**Comment by swiftix** [comment]
@swift-ci Please smoke test and merge

**Pull request merged**